### PR TITLE
Move Client4 error handling to redux middleware

### PR DESCRIPTION
**Pull request by hmhealey**
**This is a WIP.**

I started down a rabbit hole last Friday because I wanted to update TypeScript to hopefully make it easier to use subpackages outside of this repo (basically, removing [this step](https://github.com/mattermost/mattermost-webapp/tree/master/packages/types#typescript) of installing them).

Unfortunately, updating TS caused all of our try/catches to fail type checking because TS got stricter around the `unknown` type being used for the error value. I was part way through silencing those when I started to wonder why we even did error handling that way where we have to repeat that try/catch logic everywhere.

Next thing I know, I had a middleware to handle the errors, and I was going through eliminating all of that. Unfortunately, by removing the error handling right where we called the client, that caused errors to leak out in the unit tests where we didn't have middleware to catch unhandled Client4 promise rejections. I also found a bunch of cases where this meant we previously had errors occurring (due to things like loading missing profiles as a side effect of getting new posts which caused the mocking library to return an error). This lead to a whole mess of other things that just made life miserable.

I think this should pass unit tests now, but I haven't even tried using MM, so I have no idea if the E2E tests will work.

I would like to get this into master eventually, but for now, I'm going to try to pull out everything we can get into master since there's some good stuff in here even if the middleware itself (which I still think is good) is causing a bit of a quagmire.

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: stores/redux_store.jsx [hunk 1]
```diff
@@ -1,20 +1,12 @@
...
 
-export function bindActionToRedux(action, ...args) {
-    return async () => {
-        await action(...args)(store.dispatch, store.getState);
-    };
-}
-
-if (process.env.NODE_ENV !== 'production') { //eslint-disable-line no-process-env
+// eslint-disable-next-line no-process-env
+if (process.env.NODE_ENV !== 'production' || window.location.origin === 'https://community.mattermost.com') {
```
Comment (hmhealey) [inline stores/redux_store.jsx:9]: https://github.com/mattermost/mattermost-webapp/pull/10618

File: tests/test_store.tsx [hunk 2]
```diff
@@ -3,24 +3,28 @@
...
 
+import {errorHandlerMiddleware} from 'mattermost-redux/store/error_handler_middleware';
+
+import {GlobalState} from 'types/store';
+
 import {defaultIntl} from './helpers/intl-test-helper';
 
 export default function testConfigureStore(initialState = {}) {
-    return configureStore([thunk])(initialState);
+    return configureStore<GlobalState, ThunkDispatch<GlobalState, Record<string, never>, AnyAction>>([errorHandlerMiddleware, thunk])(initialState as GlobalState);
```
Comment (hmhealey) [inline tests/test_store.tsx:19]: https://github.com/mattermost/mattermost-webapp/pull/10619

**Comment by lieut-data** [comment]
Heads up that as part of our efforts to move to a monorepo, we're closing out a number of older pull requests like this one to help streamline the effort.

If you'd like to preserve these changes -- even if you're not the original author! -- feel free to resubmit the pull request against the monorepo once it's ready. You can subscribe to [mattermost-server-issue-22420](https://www.github.com/mattermost/mattermost-server/issues/22420) for status updates on this effort.

**Pull request closed without merging**